core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java

```diff
@@ -0,0 +1,77 @@
+package org.htsjdk.core.utils;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * General utilities to check and validate conditions for constructors/methods.
+ */
+public final class ValidationUtils {
+
+    // cannot be instantiated
+    private ValidationUtils() {}
+
+    /**
+     * Checks that the condition is {@code true}.
+     *
+     * @param condition required condition.
+     * @param message   the text message that would be passed to the exception thrown.
+     */
+    public static void validateArg(final boolean condition, final Supplier<String> message) {
+        if (!condition) {
+            throw new IllegalArgumentException(message.get());
+        }
+    }
+
+    /**
+     * Checks whether an index is within bounds.
+     *
+     * <p>It considers that the first index position is {@code 0}.
+     *
+     * @param index  query index.
+     * @param length collection or array size.
+     *
+     * @return same value as the input {@code index}.
+     */
+    public static int validateIndex(final int index, final int length) {
+        validateArg(index > 0, () -> "index cannot be negative: " + index);
+        validateArg(index < length, () -> "index out of bounds: " + index + ">" + (length - 1));
+        return index;
+    }
+
+    /**
+     * Checks that an {@link Object} is not {@code null}.
+     *
+     * @param object  any {@link Object}.
+     * @param message the text message that would be passed to the exception thrown.
+     * @param <T>     object type.
+     *
+     * @return the same object (if non-null).
+     *
+     * @throws IllegalArgumentException if a {@code object == null}.
+     */
+    public static <T> T nonNull(final T object, final Supplier<String> message) {

```

Comment:
I would add one arg versions of these that just a report a generic "this cannot be null" because in most places the message is not going to more useful than the stack trace.

Reply:
Done.